File: algorithms/[1094]拼车/solution.py
# topics = ["数组", "贪心算法"]

from typing import List, Tuple


class Solution:
    def carPooling(self, trips: List[List[int]], capacity: int) -> bool:
        """
        差分统计
        time O(m + n), space O(n), m 和 n 分别为数组 capacity 和数组 trips 的长度
        """
        stop: List[Tuple[int, int]] = []
        for n, s, e in trips:
            stop.append((s, n))
            stop.append((e, -n))

        # 升序
        stop.sort()

        for _, count in stop:
            capacity -= count
            if capacity < 0:
                return False

        return True

File: algorithms/[1094]拼车/solution.ts
// topics = ["数组"]

/**
 * 同 1109.航班预订统计。差分统计。
 * time O(m + n), space O(n), m 和 n 分别为数组 capacity 和数组 trips 的长度
 */
function carPooling(trips: number[][], capacity: number): boolean {
  const n = Math.max(...trips.map(item => item[2]));
  const arr = new Array<number>(n + 1).fill(0);

  for (const [cnt, start, end] of trips) {
    arr[start] += cnt;
    if (end < n) {
      arr[end] -= cnt;
    }
  }

  for (let i = 1; i <= n; i += 1) {
    arr[i] += arr[i - 1];
  }

  return capacity >= Math.max(...arr);
}
